docs(site-2): Action, TypeScript (#8588)
sveltejs/svelte: documentation/docs/05-misc/02-typescript.md

title: TypeScript
---

You can use TypeScript to type your Svelte components. The Svelte VSCode extension and the [`svelte-check` CLI](https://www.npmjs.com/package/svelte-check) will check your code for errors and provide hints about how to fix them.

## Setup

> 
> Install `typescript` and `svelte-preprocess`:

Reviewer: we should probably add a heading here and then swap the order of the vite and non-vite sections (I can't easily propose that in the UI) so that the Vite section comes first as most users will be using Vite

```suggestion

### Without Vite

Install `typescript` and `svelte-preprocess`:
```

```sh
npm install typescript svelte-preprocess --save-dev
```

> Then, create a `svelte.config.js` file at the root of your project(if it doesn't exist already) and add the following:

Reviewer: ```suggestion
Then create a `svelte.config.js` file at the root of your project (if it doesn't exist already) and add the following:
```

```ts
import preprocess from 'svelte-preprocess';

const config = {
	preprocess: preprocess()
};

export default config;
```

The line `preprocess: preprocess()` is the important one.

### Using Vite or SvelteKit

> If you're using `vite-plugin-svelte`, which is part of SvelteKit and also used in the Vite Svelte template, you can use `vitePreprocess`. It is slightly faster and does not require installing additional dependencies such as `svelte-preprocess`.

Reviewer: ```suggestion
When setting up a new SvelteKit or Svelte-flavoured Vite project, `vitePreprocess` will be added automatically if you select that you'd like to use TypeScript. It is slightly faster than `svelte-preprocess` and does not require installing additional dependencies.
```

title: TypeScript
---

> You can use TypeScript to type your Svelte components. The Svelte VSCode extension and the [`svelte-check` CLI](https://www.npmjs.com/package/svelte-check) will check your code for errors and provide hints about how to fix them.

Reviewer: ```suggestion
You can use TypeScript to type your Svelte components. The [Svelte VSCode extension](https://marketplace.visualstudio.com/items?itemName=svelte.svelte-vscode) and the [`svelte-check`](https://www.npmjs.com/package/svelte-check) CLI will check your code for errors and provide hints about how to fix them.
```

## Setup

> ### Without Vite

Reviewer: I'd put the Vite/SvelteKit section first and this one second

Install `typescript` and `svelte-preprocess`:

```sh
npm install typescript svelte-preprocess --save-dev
```

Then create a `svelte.config.js` file at the root of your project (if it doesn't exist already) and add the following:

```ts
import preprocess from 'svelte-preprocess';

const config = {
	preprocess: preprocess()
};

export default config;
```

> The line `preprocess: preprocess()` is the important one.
> 

Reviewer: ```suggestion
```

If you're using the official Vite+Svelte template, you can use the relatively faster `vitePreprocess` for faster speeds without an additional dependency. Installing `svelte-preprocess` is not required.

```ts
/// file: svelte.config.js
import { vitePreprocess } from '@sveltejs/vite-plugin-svelte';

const config = {
	preprocess: vitePreprocess()
};

export default config;
```

> If you're using SvelteKit, `vitePreprocess` is imported from `@sveltejs/kit/vite`

Reviewer: ```suggestion
If you're using SvelteKit, `vitePreprocess` can be imported from `@sveltejs/kit/vite`
```

```ts
/// file: svelte.config.js
// @noErrors
import { vitePreprocess } from '@sveltejs/kit/vite';

const config = {
	preprocess: vitePreprocess()
};

export default config;
```

## `<script lang="ts">`

To add types to your JavaScript inside the `<script>` tag, you can use the `script` tag's `lang` attribute:

```svelte
<script lang="ts">
	let name: string = 'world';

	function greet(name: string) {
		alert(`Hello, ${name}!`);
	}
</script>
```

This enables all TypeScript features inside your Svelte components.

## Props

Props can be directly typed by using the `export let` syntax:

```svelte
<script lang="ts">
	export let name: string;
</script>
```

> These are ultimately just variable declarations, so it just works.
> 

Reviewer: ```suggestion
```

> See this [guide](https://github.com/dummdidumm/rfcs/blob/ts-typedefs-within-svelte-components/text/ts-typing-props-slots-events.md#typing-props) for more information about the experimental `$$Props` interface.

## Slots

> Slots and slot props types are inferred from the type of slot props passed to them.

Reviewer: ```suggestion
Slot and slot prop types are inferred from the types of the slot props passed to them.
```

For example:

```svelte
<script lang="ts">
	export let name: string;
</script>

<slot {name} />

<!-- Later -->
<Comp let:name>
	<!--    ^ Inferred as string -->
	{name}
</Comp>
```

> See this [guide](https://github.com/dummdidumm/rfcs/blob/ts-typedefs-within-svelte-components/text/ts-typing-props-slots-events.md#typing-slots) for more information about the experimental `$$Slots` interface.

## Events

Events can be typed with `createEventDispatcher`:

```svelte
<script lang="ts">
	import { createEventDispatcher } from 'svelte';

	const dispatch = createEventDispatcher<{
		click: string;
		type: string;
	}>();

	function handleClick() {
		dispatch('click', { detail: 'hello' });
	}

	function handleType() {
		dispatch('type', { detail: 'world' });
	}
</script>

<button on:click={handleClick} on:keydown={handleType}>Click</button>
```

> However, sometimes an event dispatcher may be coming from another file/package. For example:

Reviewer: ```suggestion
However, sometimes an event dispatcher may be coming from another module. For example:
```

```svelte
<script>
	import { mixinDispatch } from 'somewhere';

	function handleClick() {
		mixinDispatch.mixinEvent('foo');
	}
</script>

<button on:click={handleClick}>Mixin</button>
```

> In this case, svelte can't understand the custom event types. So, you can use the experimental `$$Events` inerface.

Reviewer: ```suggestion
In this case, Svelte can't understand the custom event types, so you can use the experimental `$$Events` interface.
```

> See this [guide](https://github.com/dummdidumm/rfcs/blob/ts-typedefs-within-svelte-components/text/ts-typing-props-slots-events.md#typing-events) for more information about the experimental `$$Events` interface.

## Limitations

> ### No TS in Template

Reviewer: I think we've generally not used capitalization in titles, so probably more consistent to use lowercase here
```suggestion
### No TS in template
```

> You cannot use TypeScript in your template. For example, the following does not work:

Reviewer: ```suggestion
You cannot use TypeScript within the curly braces of a template's markup. For example, the following does not work:
```
